File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;
	static int N,M,size;
	static int ans = Integer.MAX_VALUE;
	static Queue<int[]> q = new ArrayDeque<>();
	static List<Chicken> ch = new ArrayList<>();
	static Chicken[] pick;
	static int[][] arr;
	
	static class Chicken {
		int r,c;
		
		Chicken(int r, int c) {
			this.r = r;
			this.c = c;
		}
	}
	
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		arr = new int[N][N];
		for(int r=0; r<N; r++) {
			st = new StringTokenizer(br.readLine());
			for(int c=0; c<N; c++) {
				arr[r][c] = Integer.parseInt(st.nextToken());
				if(arr[r][c]==1) q.add(new int[] {r+1,c+1});
				if(arr[r][c]==2) ch.add(new Chicken(r+1,c+1));
			}
		}
		size = ch.size();
		pick = new Chicken[M];
		solve(0,0);
		bw.write(ans+"");
		bw.close();
		br.close();
	}
	
	static void solve(int cnt, int idx) {
		if(cnt==M) {
			ans = Math.min(ans, calc());
			return;
		}
		for(int i=idx; i<size; i++) {
			pick[cnt]=ch.get(i);
			solve(cnt+1,i+1);
		}
	}
	
	static int calc() {
		int sum = 0;
		for(int[] house : q) {
			int dist = Integer.MAX_VALUE;
			for(Chicken c : pick) {
				dist = Math.min(dist, Math.abs(house[0]-c.r)+Math.abs(house[1]-c.c));
			}
			sum+=dist;
		}
		return sum;
	}
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <bits/stdc++.h>
using namespace std;
#define R first
#define C second
int N,M,a[51][51],vis[51][51],ret=987654321;
vector<pair<int,int>> v,cand,c;
void solve() {
    int tmp=0;
    for(int i=0; i<v.size(); i++) {
        int curR=v[i].R, curC=v[i].C, mn=987654321;
        for(int j=0; j<M; j++) {
            int dist=abs(curR-c[j].first)+abs(curC-c[j].second);
            mn=min(mn,dist);
        }
        tmp+=mn;
    }
    ret=min(ret,tmp);
}
void pick(int cnt, int idx) {
    if(cnt==M) {
        solve();
        return;
    }
    for(int i=idx; i<cand.size(); i++) {
        if(!vis[cand[i].R][cand[i].C]) {
            vis[cand[i].R][cand[i].C]=1;
            c.push_back({cand[i].R,cand[i].C});
            pick(cnt+1,i+1);
            vis[cand[i].R][cand[i].C]=0;
            c.pop_back();
        }
    }
}
int main() {
    cin >> N >> M;
    for(int r=0; r<N; r++) {
        for(int c=0; c<N; c++) {
            cin >> a[r][c];
            if(a[r][c]==1) v.push_back({r,c});
            else if(a[r][c]==2) cand.push_back({r,c});
        }
    }
    pick(0,0);
    cout << ret << '\n';
    return 0;
}